### PR TITLE
✅ amp-recaptcha-input: Wrote an integration test

### LINKED ISSUE #10059: amp-recaptcha shell extension integration test
Follow on from #9678. Once the full extension is implemented. We should do integration tests.

cc @jasti

### PR DESCRIPTION
**Pull request by torch2424**
closes #10059
relates to #20539
relates to #20540
relates to #2273

This completes the integration test for `amp-recaptcha-input`.

screen shot 2019-02-12 at 5 08 49 pm

### CODE REVIEW

File: test/integration/test-amp-recaptcha-input.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+import {Deferred} from '../../src/utils/promise';
+import {
+  createFixtureIframe,
+  poll,
+} from '../../testing/iframe';
+import {installPlatformService} from '../../src/service/platform-impl';
+import {toggleExperiment} from '../../src/experiments';
+
+
+function createFixture() {
```
Comment (lannka) [inline test/integration/test-amp-recaptcha-input.js]: nit: move helper functions to the bottom of the file

File: test/integration/test-amp-recaptcha-input.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+    return poll('Polling for hidden input', () => {
+      return fixture.doc.querySelector('input[hidden]');
+    }, undefined, 5000);
+  }).then(() => {
+    return fixture.doc.querySelector('input[hidden]');
+  });
+}
+
+describe.configure().run('amp-recaptcha-input', () => {
```
Comment (lannka) [inline test/integration/test-amp-recaptcha-input.js]: `configure()` not needed

File: build-system/recaptcha-router.js [hunk 3]
```diff
@@ -41,6 +42,7 @@ const recaptchaFrameRequestHandler = (req, res, next) => {
 };
 
 recaptchaRouter.get('/mock.js', (req, res) => {
+  console.error('Serving recaptcha mock');
```
Comment (lannka) [inline build-system/recaptcha-router.js]: remove

File: build-system/recaptcha-router.js [hunk 4]
```diff
@@ -23,6 +23,7 @@ const recaptchaRouter = require('express').Router();
 const upload = multer();
 
 const recaptchaMock = `
+console.error('I was loaded in travis!');
```
Comment (lannka) [inline build-system/recaptcha-router.js]: remove

**Comment by torch2424** [comment]
@lannka This is now good to go! 😄

File: .travis.yml [hunk 5]
```diff
@@ -36,6 +36,9 @@ addons:
   hosts:
     - ads.localhost
     - iframe.localhost
+    # CURLs amp subdomain for amp-recaptcha-input,
+    # integration tests.
+    - jgla3zmib2ggq5buc4hwi5taloh6jlvzukddfr4zltz3vay5s5rq.recaptcha.localhost
```
Comment (lannka) [inline .travis.yml]: is this domain name fixed? pls explain a bit how this string is generated

File: test/fixtures/amp-recaptcha-input.html [hunk 6]
```diff
@@ -0,0 +1,104 @@
+<!doctype html>
```
Comment (lannka) [inline test/fixtures/amp-recaptcha-input.html]: is this file still needed?

File: test/integration/test-amp-recaptcha-input.js [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+  beforeEach(() => {
+    doc = env.win.document;
+
+    const browserController = new BrowserController(env.win);
+    return browserController.waitForElementLayout('amp-recaptcha-input');
+  });
+
+  it('should be able to create the bootstrap frame', function() {
+    this.timeout(7000);
```
Comment (lannka) [inline test/integration/test-amp-recaptcha-input.js]: default timeout is 15000 now. you can remove

File: test/integration/test-amp-recaptcha-input.js [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+          .to.be.equal('amp-recaptcha');
+      expect(frame.getAttribute('name')).to.be.equal(JSON.stringify({
+        'sitekey': '6LebBGoUAAAAAHbj1oeZMBU_rze_CutlbyzpH8VE',
+        'sentinel': 'amp-recaptcha',
+      }));
+    });
+  });
+
+  it('should load the 3p recaptcha frame', function() {
+    this.timeout(7000);
```
Comment (lannka) [inline test/integration/test-amp-recaptcha-input.js]: same here

File: test/integration/test-amp-recaptcha-input.js [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+    return waitForBootstrapFrameOnLoad(doc).then(frame => {
+      expect(frame).to.be.ok;
+    });
+  });
+
+  it('should create a hidden input, ' +
```
Comment (lannka) [inline test/integration/test-amp-recaptcha-input.js]: it will be better if we can test the input kv pair is really sent over to server as body payload.

to do that, you can use request-bank.

File: .travis.yml (hunk 5 above)
Comment (torch2424) [reply .travis.yml]: Yes 😄 Remember that time I showed you this URL by running it through the `amp-toolbox` curls subdomain generator thing? Running `localhost:[whatever the karma test port is]`, gives us this URL 👍

**Comment by torch2424** [comment]
@lannka This is good to go 😄 

screen shot 2019-02-14 at 11 54 56 am

Also, opened these issues for you #20868 #20869 . That being said, Yes, using RequestBank is suuuupppeeerrr easy and handy! Great work! 👍

File: .travis.yml (hunk 5 above)
Comment (lannka) [reply .travis.yml]: please add that as comment

File: test/integration/test-amp-recaptcha-input.js [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+  });
+
+  it('should show submit-success on successful submit/response', function() {
+    return submitForm(doc).then(() => {
+      return poll('submit-success', () => {
+        return doc.querySelector('div[id="submit-success"]');
+      }, undefined, 5000);
+    });
+  });
+});
```
Comment (lannka) [inline test/integration/test-amp-recaptcha-input.js]: once you figured out how to do POST, pls add a request-bank test here

Comment (torch2424) [reply test/integration/test-amp-recaptcha-input.js]: Done

**Comment by torch2424** [comment]
Opened #21009 as a followup to this tests once Travis passes.

Also, had to change the endpoint glob for the `recaptcha.html`, as the single pass integration tests request the recaptcha html from a different path.

**Comment by torch2424** [comment]
@danielrozenberg I know I can just skip the bundle size check, but should I in this case? Says it couldn't find my v0.js?

**Comment by danielrozenberg** [comment]
Looks like the bundle-size bot had some failures last night. I'm investigating - approving this in the mean time

**Comment by torch2424** [comment]
@danielrozenberg Thank you! 😄

**Comment by rsimha** [comment]
These tests are broken on Firefox. Skipping in #21148
